front/pages/api/v1/w/[wId]/apps/[aId]/runs/index.ts

```diff
@@ -264,92 +186,194 @@ async function handler(
+            run.specification_hash = run.app_hash;
+            delete run.app_hash;
+
+            run.status.blocks = [];
+            run.results = null;
+
+            res.status(200).json({ run: run as RunType });
+          } catch (err) {
+            logger.error(
+              {
+                error: err,
+              },
+              "There was an error getting the app status."
+            );
+
+            return apiError(req, res, {
+              status_code: 400,
+              api_error: {
+                type: "run_error",
+                message: "There was an error getting the app status.",
+              },
+            });
+          }
+        })();
+      }
+
+      const usages: Usage[] = [];
+      const traces: Trace[] = [];
+
+      try {
+        // Intercept block_execution events to store token usages.
+        const parser = createParser((event) => {
+          if (event.type === "event") {
+            if (event.data) {
+              try {
+                const data = JSON.parse(event.data);
+                if (data.type === "block_execution") {
+                  if (req.body.blocking) {
+                    // Keep track of block executions for blocking requests.
+                    traces.push([
+                      [data.content.block_type, data.content.block_name],
+                      data.content.execution,
+                    ]);
+                  }
+                  const block = config[data.content.block_name];
+                  extractUsageFromExecutions(
+                    block,
+                    data.content.execution,
+                    usages
+                  );
+                }
+              } catch (err) {
+                logger.error(
+                  { error: err },
+                  "Error parsing run events while extracting usage from executions"
+                );
               }
-              return true;
-            },
-            interval: 128,
-            increment: 32,
-            maxInterval: 1024,
-            maxAttempts: 64,
-          });
-        } catch (e) {
-          return apiError(req, res, {
-            status_code: 400,
-            api_error: {
-              type: "run_error",
-              message: `There was an error polling the run status: runId=${runId} error=${e}`,
-            },
-          });
+            }
+          }
+        });
+
+        for await (const chunk of runRes.value.chunkStream) {
+          parser.feed(new TextDecoder().decode(chunk));
+          if (req.body.stream) {
+            res.write(chunk);
+            // @ts-expect-error we need to flush for streaming but TS thinks flush() does not exists.
+            res.flush();
+          }
         }
+      } catch (err) {
+        logger.error(
+          {
+            error: err,
+          },
+          "Error streaming from Dust API"
+        );
+      }
+
+      if (req.body.stream) {
+        res.end();
```

Comment:
Would be easier to read if it returned. We just need to move the run creation  usage reporting above this line?